students/elmar_m/lesson07/test_html_render.py

```diff
@@ -0,0 +1,165 @@
+        obj = hr.SelfClosingTag()
+        self.assertIsInstance(obj, hr.SelfClosingTag)
+        self.assertIsInstance(obj, hr.Element)
+        self.assertTrue(hasattr(obj, 'render'))
+    
+
+    def test_Meta(self):
+        self.assertTrue(issubclass(hr.Meta, hr.SelfClosingTag))
+        obj = hr.Meta()
+        self.assertIsInstance(obj, hr.Meta)
+        self.assertIsInstance(obj, hr.SelfClosingTag)
+        self.assertIsInstance(obj, hr.Element)
+        self.assertTrue(obj.tag == 'meta')
+        self.assertTrue(hasattr(obj, 'props'))
+        self.assertTrue(type(obj.props) == list)
+        self.assertTrue(len(obj.props) == 0)
+        self.assertTrue(hasattr(obj, 'propstring'))
+        self.assertTrue(type(obj.propstring) == str)
+        self.assertTrue(len(obj.propstring) == 0)
+
+
+    def test_Hr(self):
+        self.assertTrue(issubclass(hr.Hr, hr.SelfClosingTag))
+        obj = hr.Hr()
+        self.assertIsInstance(obj, hr.Hr)
+        self.assertIsInstance(obj, hr.SelfClosingTag)
+        self.assertIsInstance(obj, hr.Element)
+        self.assertTrue(obj.tag == 'hr')
+
+
+    def test_A(self):
+        self.assertTrue(issubclass(hr.A, hr.Element))
+        obj = hr.A('https://www.uw.edu', 'university')
+        self.assertIsInstance(obj, hr.A)
+        self.assertIsInstance(obj, hr.Element)
+        self.assertTrue(obj.tag == 'a')
+        self.assertTrue(hasattr(obj, 'render'))
+
+
+    def test_H(self):
+        self.assertTrue(issubclass(hr.H, hr.Element))
+        obj = hr.H(2, 'headertext')
+        self.assertIsInstance(obj, hr.H)
+        self.assertIsInstance(obj, hr.Element)
+        self.assertTrue(obj.tag == 'h')
+        self.assertTrue(type(obj.size) == int)
+        self.assertTrue(type(obj.text) == str)
+        self.assertTrue(hasattr(obj, 'render'))
+
+
+    def test_Ul(self):
+        self.assertTrue(issubclass(hr.Ul, hr.Element))
+        obj = hr.Ul()
+        self.assertIsInstance(obj, hr.Ul)
+        self.assertIsInstance(obj, hr.Element)
+        self.assertTrue(obj.tag == 'ul')
+        self.assertTrue(hasattr(obj, 'props'))
+        self.assertTrue(type(obj.props) == list)
+        self.assertTrue(len(obj.props) == 0)
+        self.assertTrue(hasattr(obj, 'propstring'))
+        self.assertTrue(type(obj.propstring) == str)
+        self.assertTrue(len(obj.propstring) == 0)
+
+
+    def test_Li(self):
+        self.assertTrue(issubclass(hr.Li, hr.Ul))
+        obj = hr.Li()
+        self.assertIsInstance(obj, hr.Li)
+        self.assertIsInstance(obj, hr.Ul)
+        self.assertIsInstance(obj, hr.Element)
+        self.assertTrue(obj.tag == 'li')
+        self.assertTrue(hasattr(obj, 'props'))
+        self.assertTrue(type(obj.props) == list)
+        self.assertTrue(len(obj.props) == 0)
+        self.assertTrue(hasattr(obj, 'propstring'))
+        self.assertTrue(type(obj.propstring) == str)
+        self.assertTrue(len(obj.propstring) == 0)
+
+
+if __name__ == '__main__':
+    ut.main()
+
```

Comment:
These tests need some work. Most of the assert statements are basically testing that Python's class and type systems work as expected. Although it might be a useful exercise to write these kinds of tests while you're learning, in practice you don't need them because Python itself is already very well tested so there's no need to test it yourself.

More useful tests would confirm the logic in your instance methods produces the expected results for various forms of input. So for each class that you define, you might write a test that passes in some content and asserts that the expected result is rendered as html.